db/rocks_db.go

```diff
@@ -0,0 +1,342 @@
+// +build rocksdb
+
+package db
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"runtime"
+
+	"github.com/stumble/gorocksdb"
+)
+
+func init() {
+	dbCreator := func(name string, dir string) (DB, error) {
+		return NewRocksDB(name, dir)
+	}
+	registerDBCreator(RocksDBBackend, dbCreator, false)
+}
+
+var _ DB = (*RocksDB)(nil)
+
+type RocksDB struct {
+	db     *gorocksdb.DB
+	ro     *gorocksdb.ReadOptions
+	wo     *gorocksdb.WriteOptions
+	woSync *gorocksdb.WriteOptions
+}
+
+func NewRocksDB(name string, dir string) (*RocksDB, error) {
+	// default rocksdb option, good enough for most cases, including heavy workloads.
+	// 1GB table cache, 512MB write buffer(may use 50% more on heavy workloads).
+	// compression: snappy as default, need to -lsnappy to enable.
+	bbto := gorocksdb.NewDefaultBlockBasedTableOptions()
+	bbto.SetBlockCache(gorocksdb.NewLRUCache(1 << 30))
+	bbto.SetFilterPolicy(gorocksdb.NewBloomFilter(10))
+
+	opts := gorocksdb.NewDefaultOptions()
+	opts.SetBlockBasedTableFactory(bbto)
+	opts.SetCreateIfMissing(true)
+	opts.IncreaseParallelism(runtime.NumCPU())
+	// 1.5GB maximum memory use for writebuffer.
+	opts.OptimizeLevelStyleCompaction(512 * 1024 * 1024)
+	return NewRocksDBWithOptions(name, dir, opts)
+}
+
+func NewRocksDBWithOptions(name string, dir string, opts *gorocksdb.Options) (*RocksDB, error) {
+	dbPath := filepath.Join(dir, name+".db")
+	db, err := gorocksdb.OpenDb(opts, dbPath)
+	if err != nil {
+		return nil, err
+	}
+	ro := gorocksdb.NewDefaultReadOptions()
+	wo := gorocksdb.NewDefaultWriteOptions()
+	woSync := gorocksdb.NewDefaultWriteOptions()
+	woSync.SetSync(true)
+	database := &RocksDB{
+		db:     db,
+		ro:     ro,
+		wo:     wo,
+		woSync: woSync,
+	}
+	return database, nil
+}
+
+// Implements DB.
+func (db *RocksDB) Get(key []byte) []byte {
+	key = nonNilBytes(key)
+	res, err := db.db.Get(db.ro, key)
+	if err != nil {
+		res.Free()
```

Comment:
why not put this above with `defer`? that way we'll avoid duplication s.Free() in moveSliceToBytes and here.

Reply:
turns out `res.Free()` when `err != nil` is not necessary, removed. I would prefer to have a function that moves c bytes to go slice.